### PR TITLE
suggestion for typoed crate or module

### LINKED ISSUE #76208: Typo'd Imports Result in Confusing Error Message
I wanted to use the chrono crate, so I added it to Cargo.toml and put an import at the top of main.rs: `use chono::prelude::*`, where I typo`d chrono as chono. I got this error message:

```
error[E0433]: failed to resolve: use of undeclared type or module `chono`
 --> src/main.rs:2:5
  |
2 | use chono::prelude::*;
  |     ^^^^^ use of undeclared type or module `chono`
```

I didn't see the typo, and I was confused because I thought I must have put the import in the wrong place, because it said I hadn't declared the module that I was declaring on that line. I checked the docs for E0433, and in the example given someone had forgotten to import the module at all, so I was surprised to see that error on an import line.

### Meta

`rustc --version --verbose`:
```
rustc 1.45.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.45.0
LLVM version: 10.0
```

### PR DESCRIPTION
**Pull request by TaKO8Ki**
Previously, the compiler didn't suggest similarly named crates or modules. This pull request adds a suggestion for typoed crates or modules.

#76208

before:

```
error[E0433]: failed to resolve: use of undeclared type or module `chono`
 --> src/main.rs:2:5
  |
2 | use chono::prelude::*;
  |     ^^^^^ use of undeclared type or module `chono`
```

after:

```
error[E0433]: failed to resolve: use of undeclared type or module `chono`
 --> src/main.rs:2:5
  |
2 | use chono::prelude::*;
  |     ^^^^^ 
  |     |
  |     use of undeclared crate or module `chono`
  |     help: a similar crate or module exists: `chrono`
```

### CODE REVIEW

**Comment by TaKO8Ki** [comment]
@wesleywiser 
Only [test/ui/hygiene/extern-prelude-from-opaque-fail.rs](https://github.com/rust-lang/rust/blob/6df1d82869d06b88ff413e63a1e8efbb311e3b5c/src/test/ui/hygiene/extern-prelude-from-opaque-fail.rs) failed. Could you give me some advice to solve this problem?

**Comment by wesleywiser** [comment]
I think to fully solve that issue, you'll need to do something very similar to what the identifier resolution does and traverse the scope to see what modules are currently in scope (see https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_resolve/lib.rs.html#1729-1740 which is probably what you need to call). That seems very messy though. 

@estebank do you know of a better way?

**Comment by estebank** [comment]
@wesleywiser I think that you are right that that is the "principled" approach, but I think we should land this PR as is as a stopgap (without closing the original report so we can clean this up later). What do you think?

**Comment by wesleywiser** [comment]
Oh, yes absolutely! This seems like an obvious improvement over the status quo even if there are some edge cases.

File: compiler/rustc_resolve/src/lib.rs [hunk 1]
```diff
@@ -3448,6 +3462,34 @@ impl<'a> Resolver<'a> {
...
+                    .filter(|(_, module)| {
+                        current_module.is_ancestor_of(module)
+                            && current_module.def_id() != module.def_id()
+                    })
+                    .map(|(_, module)| module.kind.name())
+                    .flatten(),
+            )
+            .filter(|c| !c.to_string().is_empty())
+            .collect::<Vec<_>>();
+        candidates.dedup();
```
[the remark this answers is not in the document]
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/lib.rs:3487]: Ok. I will fix it.

**Comment by TaKO8Ki** [comment]
@estebank @wesleywiser Thank you for your advice!

**Comment by TaKO8Ki** [comment]
@estebank So should I open a new issue about that?

File: compiler/rustc_resolve/src/lib.rs [hunk 2]
```diff
@@ -2529,7 +2530,20 @@ impl<'a> Resolver<'a> {
...
+                            (
+                                format!("use of undeclared crate or module `{}`", ident),
+                                self.find_similarly_named_module_or_crate(
+                                    ident.name,
+                                    &parent_scope.module,
+                                )
+                                .map(|sugg| {
+                                    (
+                                        vec![(ident.span, sugg.to_string())],
+                                        String::from("a similar crate or module exists"),
```
Comment (estebank) [inline compiler/rustc_resolve/src/lib.rs:2542]:
```suggestion
                                        String::from("there is a crate or module with a similar name"),
```

Ideally we would get the type of what the `ident` is from `find_similarly_named_module_or_crate` so we can be more specific in the message. We can do that later.

**Review by estebank — COMMENTED** [review]
> So should I open a new issue about that?

Please do.

File: compiler/rustc_resolve/src/lib.rs (hunk 2 above)
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/lib.rs:2542]: Ok. I'll do it in a new PR.

**Comment by TaKO8Ki** [comment]
@estebank I opened a issue about that! #89592

**Comment by estebank** [comment]
There are three tests that are failing with the same ICE:

```
------------------------------------------
stderr:
------------------------------------------
thread 'rustc' panicked at '`ModuleData::def_id` is called on a block module', compiler/rustc_resolve/src/lib.rs:607:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.57.0-nightly (86650600b 2021-10-06) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z emit-future-incompat-report -Z save-analysis -C codegen-units=1 -C prefer-dynamic -C rpath -C debuginfo=0

query stack during panic:
end of query stack

------------------------------------------



failures:
    [ui] ui/proc-macro/weird-hygiene.rs
    [ui] ui/save-analysis/issue-59134-0.rs
    [ui] ui/save-analysis/issue-59134-1.rs
```

This can likely be prevented by looking at the module kind before calling `def_id()` or using `opt_def_id()` instead.

File: compiler/rustc_resolve/src/lib.rs [hunk 3]
```diff
@@ -3482,6 +3498,35 @@ impl<'a> Resolver<'a> {
         }
         self.main_def = Some(MainDefinition { res, is_import, span });
     }
+
+    fn find_similarly_named_module_or_crate(
```
[the remark this answers is not in the document]
Comment (TaKO8Ki) [reply compiler/rustc_resolve/src/lib.rs]: I fixed that.

**Comment by TaKO8Ki** [comment]
@estebank 
> can you please commit to follow up on this soon?

Yes.

**Pull request merged**